### PR TITLE
Feature/reformat as markdown

### PR DESCRIPTION
**Pull request by kusalananda**
Initial stab at transferring the original Google Doc to GitHub and formatting it using Markdown.

Still to do: The suggestions about commit messages, from Micke. Will be done separately.

Sorry about my own commit messages here, there was too much changing to keep them useful. Basically, I'm just hoping that the text is readable.

Input wanted: Should the structure be like this, or should it be changed. I've converted some of the lists (the whole thing started as one big list) into sections and subsections where I thought it made sense.

IMOH, it doesn't look quite as nice as the original Google Doc document...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+Public interfaces include any method, function, subroutine or any
+similar interface that a user’s code may call. It also includes command
+line flags and other command line arguments that are available for the
+user to use.
+
+* If a standard way of documenting an interface is given by the language
+(e.g. JavaDoc in Java, or POD for use with `perldoc` in Perl), use that.
+* If the language does not support generation of interface documentation
+from structured inline comments or similar, you may choose to use
+something like Doxygen, or some other tool/framework that is popular for
```
Comment (nicjar) [inline README.md]: Add link to Doxygen?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+documentation of that specific language or type of code. REST APIs, for
+example, may be documented using Swagger.
```
Comment (nicjar) [inline README.md]: Add link to Swagger?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+accessible software. This also means that we will be using Git (rather
+than Subversion, Mercurial, CVS or any other code revision system) for
+keeping track of source code and documents relating to software that we
+make available on GitHub.
+
+GitHub provides excellent support for doing code reviews (more on that
+below), collecting issues (bug reports) connected with a project, and
+for organising work tasks ("projects" in GitHub speak).
+
+On GitHub, we have an Educational Account called "NBISweden"
```
Comment (nicjar) [inline README.md]: Link to https://github.com/NBISweden

File: README.md [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+other’s toes" during the development process, even if they push their
+work back to GitHub.
+* Co-developers and users have a stable master branch to use (for doing
+work in the case of the user, and as reference to their own coding in
+the case of the developer).
+* Features in "feature" branches are independent of each other. Any
+conflicts are resolved when merging.
+
+There are a host of different graphical user interfaces that helps
+keeping track of Git and the various branches in a project. SourceTree
```
Comment (nicjar) [inline README.md]: Link to https://www.sourcetreeapp.com/ (?)

File: README.md [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+is a good free one for macOS, for example.
```
Comment (nicjar) [inline README.md]: " and Windows"

File: README.md [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+* Communicate what ideas (proposed changes) you feel strongly about, and
+which ones you don’t.
+* Identify ways in which a simpler solution to the problem may be had.
+* Let the author have the last call on the final implementation,
+and move philosophical, academic or otherwise unrelated technical
+discussions to an alternate forum.
+* Seek to understand the perspective of the author.
+* Sign off the final review with a thumbs up or some other positive remark.
+
+TEXT ABOVE WILL BE EXPANDED
```
Comment (nicjar) [inline README.md]: Not needed any longer (?). Remove

File: README.md [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+
+ALL TEXT BELOW WILL EITHER BE DELETED OR MERGED INTO TEXT ABOVE
```
Comment (nicjar) [inline README.md]: Not needed any longer (?). Remove

File: README.md [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+      * [Testing](#testing)
+    * [How we use GitHub](#how-we-use-github)
+    * [How we use Git](#how-we-use-git)
+    * [General stuff about working with Git](#general-stuff-about-working-with-git)
+    * [How we do code reviews](#how-we-do-code-reviews)
+      * [Code reviewing steps](#code-reviewing-steps)
+      * [General stuff about code reviews](#general-stuff-about-code-reviews)
+
+## Things to be aware of when writing code
+
```
Comment (kusalananda) [inline README.md]: Converted this big list into subsections. I thought that the text was too long for a list.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+    Git repository file structure to avoid accidental inclusion as
+    part of the repository,
+    * Environment variables,
+    * or some sort of secured (possibly remote) storage.
+* The documentation (README/INSTALL, whichever is most appropriate)
+should mention how to instantiate those variables/files, etc.
+
+### Testing
+
+(TODO: Write me)
```
Comment (kusalananda) [inline README.md]: This is a bit of a pun, I just realised :-)

File: README.md [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+When appropriate, we use the *Git-Flow branching model*. This is a way
+of using Git branches as a help in the development cycle.
+
+For more in-depth descriptions of Git-Flow, see
+
+* [http://nvie.com/posts/a-successful-git-branching-model/](http://nvie.com/posts/a-successful-git-branching-model/)
+* [http://danielkummer.github.io/git-flow-cheatsheet/](http://danielkummer.github.io/git-flow-cheatsheet/)
+
+With Git-Flow, branches are categorised into
+
```
Comment (kusalananda) [inline README.md]: I divided the big list with its text into a tiny list plus free-standing paragraphs. Looks better, and the tiny list gives a sort of an overview.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,546 @@
+# Coding Guidelines for NBIS Developers
```
Comment (kusalananda) [inline README.md]: Markdown unfortunately does not accommodate sub-titles.

File: README.md (hunk 4 above)
Comment (kusalananda) [inline README.md]: Will do!

File: README.md (hunk 2 above)
Comment (kusalananda) [inline README.md]: Will do!

File: README.md (hunk 1 above)
Comment (kusalananda) [inline README.md]: Yes.

File: README.md (hunk 7 above)
Comment (kusalananda) [inline README.md]: I'll be moving the references into the text where appropriate.

File: README.md [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+
+* **INSTALL** (plain text or Markdown-formatted). Unless the software is
+trivial to install (e.g. just copy one file), then an INSTALL document
+should be added in which a user may find the following:
+    * External dependencies (including specific versions, where
+    applicable).
+    * Step-by-step instructions for how to install.
+    * If appropriate, how to test the installation to make sure it works.
+
+* **LICENSE** or **COPYING** (plain text). When at all possible, NBIS
```
Comment (jhagberg) [inline README.md]: I liked the layout from the google doc with item list. It was easier to read.

**Pull request merged**